File: src/IECODE.Core/Formats/Level5/G4txParser.cs
// G4TX Parser — Level-5 Graphics 4 Texture Container
// AOT-Compatible. Zero-copy: texture data is sliced from the original file buffer.

using System;
using System.Buffers;
using System.Buffers.Binary;
using System.Collections.Generic;
using System.IO;
using System.Runtime.InteropServices;
using System.Text;
using IECODE.Core.Formats.Level5.CfgBin.Tools;

namespace IECODE.Core.Formats.Level5;

/// <summary>G4TX Header (0x60 bytes).</summary>
[StructLayout(LayoutKind.Sequential, Pack = 1)]
public readonly struct G4txHeader
{
    public readonly uint Magic;
    public readonly short HeaderSize;
    public readonly short FileType;
    public readonly int Unknown1;
    public readonly int TableSize;
    public readonly ulong Zeroes1;
    public readonly ulong Zeroes2;
    public readonly short TextureCount;
    public readonly short TotalCount;
    public readonly byte Unknown2;
    public readonly byte SubTextureCount;
    public readonly short Unknown3;
    public readonly int Unknown4;
    public readonly int TextureDataSize;
    public readonly long Unknown5;
    public readonly ulong Unknown6_1;
    public readonly ulong Unknown6_2;
    public readonly ulong Unknown6_3;
    public readonly ulong Unknown6_4;
    public readonly ulong Unknown6_5;

    public bool IsValid => Magic == 0x58543447; // "G4TX" LE
}

/// <summary>G4TX Main Texture Entry (0x30 bytes).</summary>
[StructLayout(LayoutKind.Sequential, Pack = 1)]
public readonly struct G4txEntry
{
    public readonly int Unknown1;
    public readonly int NxtchOffset;
    public readonly int NxtchSize;
    public readonly int Unknown2;
    public readonly int Unknown3;
    public readonly int Unknown4;
    public readonly short Width;
    public readonly short Height;
    public readonly int Const2;
    public readonly ulong Unknown5_1;
    public readonly ulong Unknown5_2;
}

/// <summary>G4TX Sub-Texture Entry (0x18 bytes) — atlas region.</summary>
[StructLayout(LayoutKind.Sequential, Pack = 1)]
public readonly struct G4txSubEntry
{
    public readonly short EntryId;
    public readonly short Unknown1;
    public readonly short X;
    public readonly short Y;
    public readonly short Width;
    public readonly short Height;
    public readonly int Unknown2;
    public readonly int Unknown3;
    public readonly int Unknown4;
}

/// <summary>NXTCH Header (Nintendo Switch Texture Chunk).</summary>
[StructLayout(LayoutKind.Sequential, Pack = 1)]
public readonly struct NxtchHeader
{
    public readonly ulong Magic;
    public readonly int TextureDataSize;
    public readonly int Unknown1;
    public readonly int Unknown2;
    public readonly int Width;
    public readonly int Height;
    public readonly int Unknown3;
    public readonly int Unknown4;
    public readonly int Format;
    public readonly int MipMapCount;
    public readonly int TextureDataSize2;

    public bool IsValid => (Magic & 0xFFFFFFFFFF) == 0x484354584E; // "NXTCH"
}

/// <summary>Parsed G4TX texture. TextureData is a zero-copy view of the source file buffer.</summary>
public readonly record struct G4txTexture(
    byte Id,
    string Name,
    int Width,
    int Height,
    int Format,
    int MipMapCount,
    ReadOnlyMemory<byte> TextureData,
    bool IsDds,
    IReadOnlyList<G4txSubTexture> SubTextures
);

/// <summary>Parsed atlas region.</summary>
public readonly record struct G4txSubTexture(
    byte Id,
    string Name,
    short X,
    short Y,
    short Width,
    short Height
);

public static class G4txParser
{
    public const uint MAGIC = 0x47345458; // "G4TX"
    private const int HEADER_SIZE    = 0x60;
    private const int ENTRY_SIZE     = 0x30;
    private const int SUB_ENTRY_SIZE = 0x18;

    // -------------------------------------------------------------------------
    // Public API
    // -------------------------------------------------------------------------

    /// <summary>Parse all textures from a G4TX file on disk.</summary>
    public static List<G4txTexture> ParseFile(string path)
    {
        // Keep as ReadOnlyMemory<byte> so slices are zero-copy.
        ReadOnlyMemory<byte> data = File.ReadAllBytes(path);
        return ParseTextures(data);
    }

    /// <summary>
    /// Parse all textures from an in-memory buffer.
    /// <paramref name="data"/> is held alive for the lifetime of the returned textures;
    /// TextureData fields are slices of it — no extra copies.
    /// </summary>
    public static List<G4txTexture> ParseTextures(ReadOnlyMemory<byte> data)
    {
        var span   = data.Span;
        var header = ParseHeader(span);

        int entryOffset    = HEADER_SIZE;
        int subEntryOffset = entryOffset + header.TextureCount * ENTRY_SIZE;
        int hashOffset     = (subEntryOffset + header.SubTextureCount * SUB_ENTRY_SIZE + 0xF) & ~0xF;
        int idOffset       = hashOffset + header.TotalCount * 4;
        int stringOffset   = (idOffset + header.TotalCount + 0x3) & ~0x3;

        var entries    = ParseEntries(span[entryOffset..], header.TextureCount);
        var subEntries = ParseSubEntries(span[subEntryOffset..], header.SubTextureCount);
        var ids        = span.Slice(idOffset, header.TotalCount);

        // String offsets — rent from pool to avoid heap allocation.
        short[] stringOffsetsRented = ArrayPool<short>.Shared.Rent(header.TotalCount);
        try
        {
            for (int i = 0; i < header.TotalCount; i++)
                stringOffsetsRented[i] = BinaryPrimitives.ReadInt16LittleEndian(span.Slice(stringOffset + i * 2, 2));

            int nxtchBase = (header.HeaderSize + header.TableSize + 0xF) & ~0xF;
            var textures  = new List<G4txTexture>(header.TextureCount);

            for (int i = 0; i < header.TextureCount; i++)
            {
                ref readonly var entry = ref entries[i];

                string name = ReadNullTerminatedString(span[(stringOffset + stringOffsetsRented[i])..]);

                int nxtchOffset = nxtchBase + entry.NxtchOffset;

                // Zero-copy: slice the original memory buffer — no ToArray().
                ReadOnlyMemory<byte> textureData = data.Slice(nxtchOffset, entry.NxtchSize);

                bool isDds = false;
                int  width = 0, height = 0, format = 0, mipMapCount = 0;

                var texSpan = textureData.Span;
                if (texSpan.Length >= 4 && BinaryPrimitives.ReadUInt32LittleEndian(texSpan) == DdsParser.DDS_MAGIC)
                {
                    isDds = true;
                    var ddsHeader = DdsParser.ParseHeader(texSpan);
                    width       = (int)ddsHeader.Width;
                    height      = (int)ddsHeader.Height;
                    format      = (int)ddsHeader.PfFourCC;
                    mipMapCount = (int)ddsHeader.MipMapCount;
                }
                else
                {
                    var nxtch = ParseNxtchHeader(texSpan);
                    width       = nxtch.Width;
                    height      = nxtch.Height;
                    format      = nxtch.Format;
                    mipMapCount = nxtch.MipMapCount;
                }

                var subTextures = BuildSubTextures(subEntries, ids, span, stringOffset, stringOffsetsRented, i, header.TextureCount);

                textures.Add(new G4txTexture(
                    Id: ids[i],
                    Name: name,
                    Width: width,
                    Height: height,
                    Format: format,
                    MipMapCount: mipMapCount,
                    TextureData: textureData,
                    IsDds: isDds,
                    SubTextures: subTextures));
            }

            return textures;
        }
        finally
        {
            ArrayPool<short>.Shared.Return(stringOffsetsRented);
        }
    }

    /// <summary>Span overload — allocates one copy. Prefer the Memory overload when possible.</summary>
    public static List<G4txTexture> ParseTextures(ReadOnlySpan<byte> data)
        => ParseTextures(new ReadOnlyMemory<byte>(data.ToArray()));

    // -------------------------------------------------------------------------
    // NXTCH helpers
    // -------------------------------------------------------------------------

    public static NxtchHeader ParseNxtchHeader(ReadOnlySpan<byte> data)
    {
        if (data.Length < Marshal.SizeOf<NxtchHeader>())
            return default;
        return MemoryMarshal.AsRef<NxtchHeader>(data);
    }

    /// <summary>
    /// Returns the pixel data that follows the NXTCH header as a zero-copy Memory slice.
    /// </summary>
    public static ReadOnlyMemory<byte> ExtractNxtchTextureData(ReadOnlyMemory<byte> nxtchData)
    {
        int headerSize = Marshal.SizeOf<NxtchHeader>();
        return nxtchData.Length > headerSize ? nxtchData[headerSize..] : ReadOnlyMemory<byte>.Empty;
    }

    /// <summary>Span overload for callers that only have a span.</summary>
    public static ReadOnlyMemory<byte> ExtractNxtchTextureData(ReadOnlySpan<byte> nxtchData)
    {
        int headerSize = Marshal.SizeOf<NxtchHeader>();
        return nxtchData.Length > headerSize ? nxtchData[headerSize..].ToArray() : ReadOnlyMemory<byte>.Empty;
    }

    // -------------------------------------------------------------------------
    // Misc
    // -------------------------------------------------------------------------

    public static G4txHeader ParseHeader(ReadOnlySpan<byte> data)
    {
        if (data.Length < HEADER_SIZE)
            throw new InvalidDataException($"Data too small for G4TX header: {data.Length} < {HEADER_SIZE}");
        ref readonly var header = ref MemoryMarshal.AsRef<G4txHeader>(data[..HEADER_SIZE]);
        if (!header.IsValid)
            throw new InvalidDataException($"Invalid G4TX magic: 0x{header.Magic:X8}");
        return header;
    }

    public static uint CalculateNameHash(string name)
        => Crc32.Compute(Encoding.UTF8.GetBytes(name));

    public static void ExtractTextureFiles(string g4txPath, string outputDir)
    {
        Directory.CreateDirectory(outputDir);
        var textures = ParseFile(g4txPath);
        foreach (var tex in textures)
        {
            string ext  = tex.IsDds ? "dds" : "nxtch";
            string dest = Path.Combine(outputDir, $"{tex.Name}.{ext}");
            // ToArray() only at the final write boundary.
            File.WriteAllBytes(dest, tex.TextureData.ToArray());
        }
    }

    // -------------------------------------------------------------------------
    // Private helpers
    // -------------------------------------------------------------------------

    private static ReadOnlySpan<G4txEntry> ParseEntries(ReadOnlySpan<byte> data, int count)
    {
        if (data.Length < count * ENTRY_SIZE)
            throw new InvalidDataException($"Insufficient data for {count} G4TX entries");
        return MemoryMarshal.Cast<byte, G4txEntry>(data[..(count * ENTRY_SIZE)]);
    }

    private static ReadOnlySpan<G4txSubEntry> ParseSubEntries(ReadOnlySpan<byte> data, int count)
    {
        if (data.Length < count * SUB_ENTRY_SIZE)
            throw new InvalidDataException($"Insufficient data for {count} G4TX sub-entries");
        return MemoryMarshal.Cast<byte, G4txSubEntry>(data[..(count * SUB_ENTRY_SIZE)]);
    }

    private static IReadOnlyList<G4txSubTexture> BuildSubTextures(
        ReadOnlySpan<G4txSubEntry> subEntries,
        ReadOnlySpan<byte> ids,
        ReadOnlySpan<byte> fullData,
        int stringOffset,
        short[] stringOffsets,
        int textureIndex,
        int textureCount)
    {
        // Count first to avoid List resizing.
        int count = 0;
        foreach (ref readonly var sub in subEntries)
            if (sub.EntryId == textureIndex) count++;

        if (count == 0)
            return Array.Empty<G4txSubTexture>();

        var result   = new G4txSubTexture[count];
        int written  = 0;
        int subIdBase = textureCount; // sub-texture IDs start after main textures
        int subIdx   = 0;

        foreach (ref readonly var sub in subEntries)
        {
            if (sub.EntryId == textureIndex)
            {
                int absoluteId = subIdBase + subIdx;
                string subName = ReadNullTerminatedString(fullData[(stringOffset + stringOffsets[absoluteId])..]);
                result[written++] = new G4txSubTexture(
                    Id: ids[absoluteId],
                    Name: subName,
                    X: sub.X, Y: sub.Y,
                    Width: sub.Width, Height: sub.Height);
            }
            subIdx++;
        }

        return result;
    }

    private static string ReadNullTerminatedString(ReadOnlySpan<byte> data)
    {
        int end = data.IndexOf((byte)0);
        if (end < 0) end = data.Length;
        return Encoding.UTF8.GetString(data[..end]);
    }
}

File: packages/inagle/src/parsers/binary/g4tx-parser.ts
/**
 * @file g4tx-parser.ts
 * @description Pure TypeScript parser for Level-5 G4TX texture containers
 * Ported from IECODE.Core/Formats/Level5/G4txParser.cs
 *
 * G4TX (Graphics 4 Texture) is Level-5's texture container format.
 * It wraps NXTCH (Nintendo Switch) or DDS textures with metadata.
 *
 * Format Structure:
 * - Header (0x60 bytes): Magic "G4TX" + metadata
 * - Entry table: Main texture entries (0x30 bytes each)
 * - Sub-entry table: Atlas regions (0x18 bytes each)
 * - Hash table: CRC32 hashes for texture names
 * - ID table: Texture IDs
 * - String table: Null-terminated texture names
 * - Texture data: NXTCH or DDS chunks
 */

// ============================================================================
// Types
// ============================================================================

export interface G4txHeader {
	magic: number; // "G4TX" (0x47345458)
	headerSize: number; // 0x60
	fileType: number; // 0x65
	tableSize: number; // Size of entry tables
	textureCount: number; // Number of main textures
	totalCount: number; // textureCount + subTextureCount
	subTextureCount: number; // Number of sub-textures (atlas regions)
	textureDataSize: number; // Total size of NXTCH data
}

export interface G4txEntry {
	nxtchOffset: number; // Offset to NXTCH chunk
	nxtchSize: number; // Size of NXTCH chunk
	width: number; // Texture width
	height: number; // Texture height
}

export interface G4txSubEntry {
	entryId: number; // Parent texture index
	x: number; // Region X offset
	y: number; // Region Y offset
	width: number; // Region width
	height: number; // Region height
}

export interface NxtchHeader {
	magic: bigint; // "NXTCH000"
	textureDataSize: number;
	width: number;
	height: number;
	format: number; // Texture format (BC7, BC1, etc.)
	mipMapCount: number;
	isValid: boolean;
}

export interface G4txSubTexture {
	id: number;
	name: string;
	x: number;
	y: number;
	width: number;
	height: number;
}

export interface G4txTexture {
	id: number;
	name: string;
	width: number;
	height: number;
	format: number;
	mipMapCount: number;
	textureData: Uint8Array;
	isDds: boolean;
	subTextures: G4txSubTexture[];
}

/**
 * NXTCH texture format enumeration
 */
export enum NxtchTextureFormat {
	Unknown = 0,
	BC1 = 0x01, // DXT1
	BC2 = 0x02, // DXT3
	BC3 = 0x03, // DXT5
	BC4 = 0x04, // ATI1/BC4
	BC5 = 0x05, // ATI2/BC5
	BC6 = 0x06, // BC6H
	BC7 = 0x07, // BC7
	RGBA8 = 0x1f, // RGBA8888
}

// ============================================================================
// Constants
// ============================================================================

export const G4TX_MAGIC = 0x58543447; // "G4TX" little-endian
export const DDS_MAGIC = 0x20534444; // "DDS " little-endian
const HEADER_SIZE = 0x60;
const ENTRY_SIZE = 0x30;
const SUB_ENTRY_SIZE = 0x18;
const NXTCH_HEADER_SIZE = 44;
const NXTCH_MAGIC_PREFIX = 0x484354584e; // "NXTCH" (5 bytes)

// ============================================================================
// Parser Implementation
// ============================================================================

/**
 * Check if data is a valid G4TX file.
 */
export function isG4tx(data: Uint8Array): boolean {
	if (data.length < 4) return false;
	const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
	return view.getUint32(0, true) === G4TX_MAGIC;
}

/**
 * Parse G4TX header from binary data.
 */
export function parseHeader(data: Uint8Array): G4txHeader {
	if (data.length < HEADER_SIZE) {
		throw new Error(`Data too small for G4TX header: ${data.length} < ${HEADER_SIZE}`);
	}

	const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
	const magic = view.getUint32(0, true);

	if (magic !== G4TX_MAGIC) {
		throw new Error(`Invalid G4TX magic: 0x${magic.toString(16).toUpperCase()}`);
	}

	return {
		magic,
		headerSize: view.getInt16(4, true),
		fileType: view.getInt16(6, true),
		tableSize: view.getInt32(12, true),
		textureCount: view.getInt16(0x20, true),
		totalCount: view.getInt16(0x22, true),
		subTextureCount: view.getUint8(0x25),
		textureDataSize: view.getInt32(0x2c, true),
	};
}

/**
 * Parse NXTCH header from texture chunk data.
 */
export function parseNxtchHeader(data: Uint8Array): NxtchHeader {
	if (data.length < NXTCH_HEADER_SIZE) {
		return {
			magic: 0n,
			textureDataSize: 0,
			width: 0,
			height: 0,
			format: 0,
			mipMapCount: 0,
			isValid: false,
		};
	}

	const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
	const magic = view.getBigUint64(0, true);

	// Check if first 5 bytes match "NXTCH"
	const isValid = (magic & 0xffffffffffn) === BigInt(NXTCH_MAGIC_PREFIX);

	return {
		magic,
		textureDataSize: view.getInt32(8, true),
		width: view.getInt32(20, true),
		height: view.getInt32(24, true),
		format: view.getInt32(32, true),
		mipMapCount: view.getInt32(36, true),
		isValid,
	};
}

/**
 * Parse all textures from G4TX file.
 */
export function parseTextures(data: Uint8Array): G4txTexture[] {
	const header = parseHeader(data);
	const textures: G4txTexture[] = [];
	const view = new DataView(data.buffer, data.byteOffset, data.byteLength);

	// Calculate offsets
	const entryOffset = HEADER_SIZE;
	const subEntryOffset = entryOffset + header.textureCount * ENTRY_SIZE;
	const hashOffset = alignUp(subEntryOffset + header.subTextureCount * SUB_ENTRY_SIZE, 0x10);
	const idOffset = hashOffset + header.totalCount * 4;
	const stringOffset = alignUp(idOffset + header.totalCount, 4);

	// Parse entries
	const entries: G4txEntry[] = [];
	for (let i = 0; i < header.textureCount; i++) {
		const offset = entryOffset + i * ENTRY_SIZE;
		entries.push({
			nxtchOffset: view.getInt32(offset + 4, true),
			nxtchSize: view.getInt32(offset + 8, true),
			width: view.getInt16(offset + 24, true),
			height: view.getInt16(offset + 26, true),
		});
	}

	// Parse sub-entries
	const subEntries: G4txSubEntry[] = [];
	for (let i = 0; i < header.subTextureCount; i++) {
		const offset = subEntryOffset + i * SUB_ENTRY_SIZE;
		subEntries.push({
			entryId: view.getInt16(offset, true),
			x: view.getInt16(offset + 4, true),
			y: view.getInt16(offset + 6, true),
			width: view.getInt16(offset + 8, true),
			height: view.getInt16(offset + 10, true),
		});
	}

	// Read IDs
	const ids = data.slice(idOffset, idOffset + header.totalCount);

	// Read string offsets
	const stringOffsets: number[] = [];
	for (let i = 0; i < header.totalCount; i++) {
		stringOffsets.push(view.getInt16(stringOffset + i * 2, true));
	}

	// Calculate NXTCH base offset
	const nxtchBase = alignUp(header.headerSize + header.tableSize, 0x10);

	// Parse each main texture
	for (let i = 0; i < header.textureCount; i++) {
		const entry = entries[i];

		// Read texture name
		const name = readNullTerminatedString(data, stringOffset + stringOffsets[i]);

		// Extract texture data
		const nxtchOffset = nxtchBase + entry.nxtchOffset;
		const textureData = data.slice(nxtchOffset, nxtchOffset + entry.nxtchSize);

		// Check for DDS Magic (DX11/PC)
		let isDds = false;
		let width = 0;
		let height = 0;
		let format = 0;
		let mipMapCount = 0;

		if (textureData.length >= 4) {
			const texView = new DataView(
				textureData.buffer,
				textureData.byteOffset,
				textureData.byteLength
			);
			const texMagic = texView.getUint32(0, true);

			if (texMagic === DDS_MAGIC) {
				isDds = true;
				// Parse DDS header
				width = texView.getInt32(16, true);
				height = texView.getInt32(12, true);
				mipMapCount = texView.getInt32(28, true);
				format = texView.getUint32(84, true); // FourCC
			} else {
				// Parse NXTCH header
				const nxtchHeader = parseNxtchHeader(textureData);
				width = nxtchHeader.width;
				height = nxtchHeader.height;
				format = nxtchHeader.format;
				mipMapCount = nxtchHeader.mipMapCount;
			}
		}

		// Find sub-textures for this entry
		const subTextures: G4txSubTexture[] = [];
		let subEntryId = header.textureCount;

		for (const subEntry of subEntries) {
			if (subEntry.entryId === i) {
				const subName = readNullTerminatedString(data, stringOffset + stringOffsets[subEntryId]);
				subTextures.push({
					id: ids[subEntryId],
					name: subName,
					x: subEntry.x,
					y: subEntry.y,
					width: subEntry.width,
					height: subEntry.height,
				});
				subEntryId++;
			}
		}

		textures.push({
			id: ids[i],
			name,
			width,
			height,
			format,
			mipMapCount,
			textureData,
			isDds,
			subTextures,
		});
	}

	return textures;
}

/**
 * Get texture format name from format code.
 */
export function getFormatName(format: number): string {
	switch (format) {
		case NxtchTextureFormat.BC1:
			return "BC1/DXT1";
		case NxtchTextureFormat.BC2:
			return "BC2/DXT3";
		case NxtchTextureFormat.BC3:
			return "BC3/DXT5";
		case NxtchTextureFormat.BC4:
			return "BC4/ATI1";
		case NxtchTextureFormat.BC5:
			return "BC5/ATI2";
		case NxtchTextureFormat.BC6:
			return "BC6H";
		case NxtchTextureFormat.BC7:
			return "BC7";
		case NxtchTextureFormat.RGBA8:
			return "RGBA8";
		default:
			return `Unknown(0x${format.toString(16)})`;
	}
}

/**
 * Convert NXTCH to DDS format.
 */
export function convertNxtchToDds(nxtchData: Uint8Array): Uint8Array {
	const nxtchHeader = parseNxtchHeader(nxtchData);
	if (!nxtchHeader.isValid) {
		throw new Error("Invalid NXTCH header");
	}

	const textureData = nxtchData.slice(NXTCH_HEADER_SIZE);

	// Create DDS header
	const ddsHeader = createDdsHeader(nxtchHeader);

	// Combine header + texture data
	const ddsData = new Uint8Array(ddsHeader.length + textureData.length);
	ddsData.set(ddsHeader, 0);
	ddsData.set(textureData, ddsHeader.length);

	return ddsData;
}

function createDdsHeader(nxtchHeader: NxtchHeader): Uint8Array {
	const ddsHeader = new Uint8Array(128);
	const view = new DataView(ddsHeader.buffer);

	// DDS magic
	view.setUint32(0, DDS_MAGIC, true);

	// DDS_HEADER
	view.setInt32(4, 124, true); // dwSize
	view.setUint32(8, 0x1007, true); // dwFlags
	view.setInt32(12, nxtchHeader.height, true);
	view.setInt32(16, nxtchHeader.width, true);

	// Calculate pitch
	const isPitchBC1 =
		nxtchHeader.format === NxtchTextureFormat.BC1 || nxtchHeader.format === NxtchTextureFormat.BC4;
	const pitchOrLinearSize = isPitchBC1
		? Math.max(1, Math.floor((nxtchHeader.width + 3) / 4)) * 8
		: Math.max(1, Math.floor((nxtchHeader.width + 3) / 4)) * 16;

	view.setInt32(20, pitchOrLinearSize, true);
	view.setInt32(28, Math.max(1, nxtchHeader.mipMapCount), true);

	// DDS_PIXELFORMAT
	view.setInt32(76, 32, true); // dwSize
	view.setUint32(80, 0x4, true); // dwFlags (FOURCC)

	// FourCC based on format
	let fourcc: number;
	switch (nxtchHeader.format) {
		case NxtchTextureFormat.BC1:
			fourcc = 0x31545844; // "DXT1"
			break;
		case NxtchTextureFormat.BC2:
			fourcc = 0x33545844; // "DXT3"
			break;
		case NxtchTextureFormat.BC3:
			fourcc = 0x35545844; // "DXT5"
			break;
		case NxtchTextureFormat.BC4:
			fourcc = 0x31495441; // "ATI1"
			break;
		case NxtchTextureFormat.BC5:
			fourcc = 0x32495441; // "ATI2"
			break;
		default:
			fourcc = 0x00000000; // DX10 header needed
			break;
	}
	view.setUint32(84, fourcc, true);

	// DDS_HEADER.dwCaps
	view.setUint32(108, 0x1000, true); // TEXTURE

	return ddsHeader;
}

// ============================================================================
// Utility Functions
// ============================================================================

function readNullTerminatedString(data: Uint8Array, offset: number): string {
	let end = offset;
	while (end < data.length && data[end] !== 0) {
		end++;
	}
	const decoder = new TextDecoder("utf-8");
	return decoder.decode(data.slice(offset, end));
}

function alignUp(value: number, alignment: number): number {
	return Math.ceil(value / alignment) * alignment;
}

// ============================================================================
// High-Level API
// ============================================================================

export interface G4txInfo {
	textureCount: number;
	subTextureCount: number;
	textures: {
		name: string;
		width: number;
		height: number;
		format: string;
		isDds: boolean;
		subTextures: {
			name: string;
			x: number;
			y: number;
			width: number;
			height: number;
		}[];
	}[];
}

/**
 * Parse G4TX and return summary info.
 */
export function getG4txInfo(data: Uint8Array): G4txInfo {
	const textures = parseTextures(data);
	const header = parseHeader(data);

	return {
		textureCount: header.textureCount,
		subTextureCount: header.subTextureCount,
		textures: textures.map((t) => ({
			name: t.name,
			width: t.width,
			height: t.height,
			format: getFormatName(t.format),
			isDds: t.isDds,
			subTextures: t.subTextures.map((s) => ({
				name: s.name,
				x: s.x,
				y: s.y,
				width: s.width,
				height: s.height,
			})),
		})),
	};
}
